src/kubernetes/api_watcher.rs

```diff
@@ -0,0 +1,160 @@
+//! A watcher based on the k8s API.
+
+use super::{
+    client::Client,
+    stream as k8s_stream,
+    watch_request_builder::WatchRequestBuilder,
+    watcher::{self, Watcher},
+};
+use futures::{
+    future::BoxFuture,
+    stream::{BoxStream, Stream},
+};
+use http02::StatusCode;
+use hyper13::Error as BodyError;
+use k8s_openapi::{WatchOptional, WatchResponse};
+use snafu::{ResultExt, Snafu};
+
+/// A simple watcher atop of the Kubernetes API [`Client`].
+pub struct ApiWatcher<B>
+where
+    B: 'static,
+{
+    client: Client,
+    request_builder: B,
+}
+
+impl<B> ApiWatcher<B>
+where
+    B: 'static,
+{
+    /// Create a new [`ApiWatcher`].
+    pub fn new(client: Client, request_builder: B) -> Self {
+        Self {
+            client,
+            request_builder,
+        }
+    }
+}
+
+impl<B> ApiWatcher<B>
+where
+    B: 'static + WatchRequestBuilder,
+    <B as WatchRequestBuilder>::Object: Send + Unpin,
+{
+    async fn invoke(
+        &mut self,
+        watch_optional: WatchOptional<'_>,
+    ) -> Result<
+        impl Stream<
+                Item = Result<
+                    WatchResponse<<B as WatchRequestBuilder>::Object>,
+                    k8s_stream::Error<BodyError>,
+                >,
+            > + 'static,
+        watcher::invocation::Error<invocation::Error>,
+    > {
+        // Prepare request.
+        let request = self
+            .request_builder
+            .build(watch_optional)
+            .context(invocation::RequestPreparation)?;
+        trace!(message = "request prepared", ?request);
```

Comment:
Please make this an internal event.  You can do something like this https://github.com/timberio/vector/blob/1d88625ce5c35708a42d612c7abc6b4087676f78/src/internal_events/wasm/event_processing.rs to track state and timings if you need. :)

Reply:
This is cool, I'll try to adopt a similar approach here!

Reply:
I noticed you're using a `vector-wasm` crate. Would it be a good idea to move `kubernetes` mod to its own crate in a similar fashion? :thinking: 
It doesn't really depend on anything from the vector besides the HTTP facilities, but those can abstracted away easily.

Reply:
Don't we want every HTTP call instrumented btw? I've added very little logging myself because there's already plenty at the lower levels.
In this sense, preparing request is not as important to emit as an internal event, imo, as other parts of it's lifetime: when it gets picked up for execution by the HTTP client and etc.

Initially, it seemed like a good idea to convert this into an event - but then, is this particular event interesting enough? Wouldn't it make more sense to emit events from the `kubernetes::client`?